File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(nums);

        for(int i=0; i<nums.length; i++){
            if(i>0 && nums[i]==nums[i-1]){
                continue;
            }

            int j = i+1;
            int k = nums.length - 1;

            while(j<k){
                if(nums[i]+nums[j]+nums[k]==0){
                    res.add(Arrays.asList(nums[i], nums[j], nums[k]));
                    
                    while(j<k && nums[j]==nums[j+1]){
                        j++;
                    }

                    while(j<k && nums[k]==nums[k-1]){
                        k--;
                    }
                }

                if(nums[j]+nums[k]>(-1)*nums[i]){
                    k--;
                }
                else{
                    j++;
                }
            }
        }

        return res;
    }
}

File: 0015-3sum/0015-3sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        n = len(nums)
        nums.sort()
        res = []

        for curr in range(0, n):
            low = curr+1
            high = n-1

            if curr > 0 and nums[curr-1] == nums[curr]:
                continue
            

            while low<high:
                if nums[low] + nums[high] + nums[curr] == 0:
                    res.append([nums[curr], nums[low] , nums[high]])
                    while low < high and nums[low+1] == nums[low]:
                        low+=1
                    while low<high and nums[high-1] == nums[high]:
                        high-=1

                if nums[low] + nums[high] > (-1) * nums[curr]:
                    high -= 1
                else:
                    low += 1


        return res
